src/passes/OptimizeInstructions.cpp

```diff
@@ -1738,16 +1743,105 @@ struct OptimizeInstructions
     return nullptr;
   }
 
+  // (x op C1) + (x op C2)   ==>   x * (C1' + C2')
+  // (x op C1) - (x op C2)   ==>   x * (C1' - C2')
+  // where op = (`*`|`<<`)
+  // x * y + x * z   ==>   (y + z) * x
+  // x * y - x * z   ==>   (y - z) * x
+  Expression* optimizeBinaryDistributions(Binary* curr) {
+    using namespace Match;
+    assert(curr->type.isInteger());
+
+    Expression* ll = nullptr;
+    Expression* rl = nullptr;
+    Expression* lr = nullptr;
+    Expression* rr = nullptr;
+    Binary* left = nullptr;
+    Binary* right = nullptr;
+    BinaryOp op;
+
+    // canonicalize
+    // (x << C1) op (x << C2)
+    // to
+    // (x * (1 << C1)) op (x * (1 << C2))
+    if (matches(curr,
+                binary(&op,
+                       binary(&left, any(&ll), any(&lr)),
+                       binary(&right, pure(&rl), any(&rr)))) &&
+        (op == Abstract::getBinary(curr->type, Abstract::Add) ||
+         op == Abstract::getBinary(curr->type, Abstract::Sub)) &&
+        ExpressionAnalyzer::equal(ll, rl)) {
+      Const *c1, *c2;
+      if (matches(left, binary(Abstract::Shl, any(), constant(&c1)))) {
+        left->op = Abstract::getBinary(left->type, Abstract::Mul);
+        c1->value = Literal::makeFromInt32(1, left->type).shl(c1->value);
+      }
+      if (matches(right, binary(Abstract::Shl, any(), constant(&c2)))) {
+        right->op = Abstract::getBinary(right->type, Abstract::Mul);
+        c2->value = Literal::makeFromInt32(1, right->type).shl(c2->value);
+      }
+    }
+
+    if (matches(curr,
+                binary(&op,
+                       binary(&left, any(&ll), any(&lr)),
+                       binary(&right, pure(&rl), pure(&rr)))) &&
+        (op == Abstract::getBinary(curr->type, Abstract::Add) ||
+         op == Abstract::getBinary(curr->type, Abstract::Sub))) {
```

Comment:
@tlively I think something composible for several operations in one place. This could be more clearer and faster.
```cpp
    if (matches(curr,
                binary(or(Abstract::Add, Abstract::Sub),
                       binary(&left, any(&ll), any(&lr)),
                       binary(&right, pure(&rl), pure(&rr)))))
```
Could we have something like that? wdyt?

Reply:
because currently we should match a lot of binary expressions and check their side effects which pretty expensive. But if we could narrow the search down to only two operations (`add` and `sub` for current) and do it as early as possible, then it would be much faster and more accurate.

Reply:
Hmm, yeah, I think that makes sense after all. Give me a little time and I can try to implement something nice for that.

Reply:
awesome!